File: Test/Bench/go/benchmark.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"time"

	benchperf "protobuf-lean-benchmark/generated"

	"google.golang.org/protobuf/proto"
)

var protobufVersion = "unknown"

const (
	fnvOffset uint64 = 14695981039346656037
	fnvPrime  uint64 = 1099511628211
)

type result struct {
	dataSetupNS  uint64
	inputSetupNS uint64
	firstNS      uint64
	steadyNS     uint64
	outputBytes  uint64
	contentHash  uint64
	outputHash   uint64
	checksum     uint64
}

func hashByte(hash uint64, value byte) uint64 {
	return (hash ^ uint64(value)) * fnvPrime
}

func hashU64(hash, value uint64) uint64 {
	for range 8 {
		hash = hashByte(hash, byte(value))
		value >>= 8
	}
	return hash
}

func hashBytesWithLength(hash uint64, bytes []byte) uint64 {
	hash = hashU64(hash, uint64(len(bytes)))
	for _, value := range bytes {
		hash = hashByte(hash, value)
	}
	return hash
}

func hashBytes(bytes []byte) uint64 {
	hash := fnvOffset
	for _, value := range bytes {
		hash = hashByte(hash, value)
	}
	return hash
}

func contentHash(batch *benchperf.Batch) uint64 {
	hash := hashBytesWithLength(fnvOffset, []byte(batch.GetLabel()))
	hash = hashU64(hash, uint64(len(batch.GetItems())))
	for _, item := range batch.GetItems() {
		hash = hashU64(hash, uint64(item.GetId()))
		hash = hashBytesWithLength(hash, []byte(item.GetName()))
		hash = hashU64(hash, uint64(len(item.GetScores())))
		for _, score := range item.GetScores() {
			hash = hashU64(hash, uint64(uint32(score)))
		}
		hash = hashBytesWithLength(hash, item.GetPayload())
		if metadata := item.GetMeta(); metadata != nil {
			hash = hashByte(hash, 1)
			hash = hashBytesWithLength(hash, []byte(metadata.GetSource()))
			hash = hashU64(hash, metadata.GetCreatedAt())
			if metadata.GetActive() {
				hash = hashByte(hash, 1)
			} else {
				hash = hashByte(hash, 0)
			}
		} else {
			hash = hashByte(hash, 0)
		}
		hash = hashU64(hash, uint64(len(item.GetTags())))
		for _, tag := range item.GetTags() {
			hash = hashBytesWithLength(hash, []byte(tag))
		}
		hash = hashBytesWithLength(hash, []byte(item.GetNote()))
	}
	return hash
}

func makeBatch(count uint64) (*benchperf.Batch, error) {
	if count > uint64(^uint(0)>>1) {
		return nil, fmt.Errorf("item count is too large for the Go runtime")
	}
	batch := &benchperf.Batch{
		Label: fmt.Sprintf("batch-%d", count),
		Items: make([]*benchperf.Item, 0, int(count)),
	}
	for i := uint64(0); i < count; i++ {
		payload := make([]byte, 48+i%16)
		for j := range payload {
			payload[j] = byte((i*31 + uint64(j)*17 + 13) % 251)
		}
		item := &benchperf.Item{
			Id:   uint32(i),
			Name: fmt.Sprintf("item-%d", i),
			Scores: []int32{
				int32((i+1)*3 - 19), int32((i+1)*4 - 19),
				int32((i+1)*5 - 19), int32((i+1)*6 - 19),
				int32((i+1)*7 - 19), int32((i+1)*8 - 19),
				int32((i+1)*9 - 19), int32((i+1)*10 - 19),
			},
			Payload: payload,
			Meta: &benchperf.Meta{
				Source:    fmt.Sprintf("source-%d", i%11),
				CreatedAt: 1700000000 + i*17,
				Active:    i%2 == 0,
			},
			Tags: []string{
				fmt.Sprintf("tag-%d", i%5),
				fmt.Sprintf("group-%d", i%9),
				fmt.Sprintf("bucket-%d", i%13),
				fmt.Sprintf("region-%d", i%7),
			},
			Note: fmt.Sprintf("note-%d-%d", i%17, i*3),
		}
		batch.Items = append(batch.Items, item)
	}
	return batch, nil
}

func encode(batch *benchperf.Batch) ([]byte, error) {
	return proto.Marshal(batch)
}

func decode(bytes []byte) (*benchperf.Batch, error) {
	batch := &benchperf.Batch{}
	if err := proto.Unmarshal(bytes, batch); err != nil {
		return nil, err
	}
	return batch, nil
}

func consumeBytes(bytes []byte) uint64 {
	if len(bytes) == 0 {
		return 0
	}
	return uint64(len(bytes)) + uint64(bytes[0]) + uint64(bytes[len(bytes)-1])
}

func consumeBatch(batch *benchperf.Batch) uint64 {
	items := batch.GetItems()
	if len(items) == 0 {
		return uint64(len(batch.GetLabel()))
	}
	return uint64(len(items)) + uint64(items[0].GetId()) +
		uint64(items[len(items)-1].GetId()) + uint64(len(batch.GetLabel()))
}

func elapsedNS(start, stop time.Time) uint64 {
	return uint64(stop.Sub(start).Nanoseconds())
}

func runEncode(items, iterations uint64, validate bool) (result, error) {
	var output result
	start := time.Now()
	batch, err := makeBatch(items)
	if err != nil {
		return output, err
	}
	output.dataSetupNS = elapsedNS(start, time.Now())
	output.contentHash = contentHash(batch)

	start = time.Now()
	last, err := encode(batch)
	if err != nil {
		return output, err
	}
	output.firstNS = elapsedNS(start, time.Now())
	output.checksum = consumeBytes(last)

	start = time.Now()
	for range iterations {
		bytes, err := encode(batch)
		if err != nil {
			return output, err
		}
		output.checksum += consumeBytes(bytes)
		last = bytes
	}
	output.steadyNS = elapsedNS(start, time.Now())
	if validate {
		decoded, err := decode(last)
		if err != nil {
			return output, err
		}
		if actual := contentHash(decoded); actual != output.contentHash {
			return output, fmt.Errorf("go-binary encode content mismatch: expected %d, got %d", output.contentHash, actual)
		}
	}
	output.outputBytes = uint64(len(last))
	output.outputHash = hashBytes(last)
	return output, nil
}

func runDecode(items, iterations uint64, validate bool) (result, error) {
	var output result
	start := time.Now()
	batch, err := makeBatch(items)
	if err != nil {
		return output, err
	}
	output.dataSetupNS = elapsedNS(start, time.Now())
	output.contentHash = contentHash(batch)

	start = time.Now()
	input, err := encode(batch)
	if err != nil {
		return output, err
	}
	output.inputSetupNS = elapsedNS(start, time.Now())

	start = time.Now()
	last, err := decode(input)
	if err != nil {
		return output, err
	}
	output.firstNS = elapsedNS(start, time.Now())
	output.checksum = consumeBatch(last)

	start = time.Now()
	for range iterations {
		decoded, err := decode(input)
		if err != nil {
			return output, err
		}
		output.checksum += consumeBatch(decoded)
		last = decoded
	}
	output.steadyNS = elapsedNS(start, time.Now())
	if validate {
		if actual := contentHash(last); actual != output.contentHash {
			return output, fmt.Errorf("go-binary decode content mismatch: expected %d, got %d", output.contentHash, actual)
		}
	}
	output.outputBytes = uint64(len(input))
	output.outputHash = hashBytes(input)
	return output, nil
}

func parseU64(name, value string) (uint64, error) {
	parsed, err := strconv.ParseUint(value, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid %s: %w", name, err)
	}
	return parsed, nil
}

func printResult(operation string, items, iterations uint64, validate bool, value result) {
	perOperation := uint64(0)
	if iterations != 0 {
		perOperation = value.steadyNS / iterations
	}
	validation := 0
	if validate {
		validation = 1
	}
	fmt.Printf("BENCH_RESULT implementation=go-binary operation=%s items=%d iterations=%d data_setup_ns=%d input_setup_ns=%d first_ns=%d steady_ns=%d steady_ns_per_op=%d output_bytes=%d content_hash=%d output_hash=%d checksum=%d validation=%d runtime_version=%s protobuf_version=%s\n",
		operation, items, iterations, value.dataSetupNS, value.inputSetupNS,
		value.firstNS, value.steadyNS, perOperation, value.outputBytes,
		value.contentHash, value.outputHash, value.checksum, validation,
		runtime.Version(), protobufVersion)
}

func main() {
	if len(os.Args) == 2 && os.Args[1] == "startup" {
		fmt.Printf("BENCH_RESULT implementation=go-runtime operation=startup items=0 iterations=0 data_setup_ns=0 input_setup_ns=0 first_ns=0 steady_ns=0 steady_ns_per_op=0 output_bytes=0 content_hash=0 output_hash=0 checksum=0 validation=0 runtime_version=%s protobuf_version=%s\n", runtime.Version(), protobufVersion)
		return
	}
	if len(os.Args) == 2 && os.Args[1] == "version" {
		fmt.Printf("%s %s\n", protobufVersion, runtime.Version())
		return
	}
	if len(os.Args) != 5 {
		fmt.Fprintln(os.Stderr, "usage: <encode|decode> <items> <steady-iterations> <validate:0|1>")
		os.Exit(2)
	}
	operation := os.Args[1]
	items, err := parseU64("items", os.Args[2])
	if err != nil {
		fmt.Fprintln(os.Stderr, "benchmark error:", err)
		os.Exit(2)
	}
	iterations, err := parseU64("steady-iterations", os.Args[3])
	if err != nil {
		fmt.Fprintln(os.Stderr, "benchmark error:", err)
		os.Exit(2)
	}
	validation, err := parseU64("validate", os.Args[4])
	if err != nil || validation > 1 {
		fmt.Fprintln(os.Stderr, "benchmark error: validate must be 0 or 1")
		os.Exit(2)
	}
	validate := validation == 1
	var output result
	if operation == "encode" {
		output, err = runEncode(items, iterations, validate)
	} else if operation == "decode" {
		output, err = runDecode(items, iterations, validate)
	} else {
		err = fmt.Errorf("operation must be encode or decode")
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, "benchmark error:", err)
		os.Exit(2)
	}
	printResult(operation, items, iterations, validate, output)
}

File: Test/Bench/cpp/benchmark.cc
#include "Perf.pb.h"

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <string_view>
#include <utility>

namespace {

using Batch = bench::perf::Batch;
using Clock = std::chrono::steady_clock;

constexpr std::uint64_t kFnvOffset = 14695981039346656037ULL;
constexpr std::uint64_t kFnvPrime = 1099511628211ULL;

std::uint64_t hash_byte(std::uint64_t hash, std::uint8_t byte) {
  return (hash ^ byte) * kFnvPrime;
}

std::uint64_t hash_u64(std::uint64_t hash, std::uint64_t value) {
  for (int i = 0; i < 8; ++i) {
    hash = hash_byte(hash, static_cast<std::uint8_t>(value));
    value >>= 8;
  }
  return hash;
}

std::uint64_t hash_bytes_with_length(std::uint64_t hash,
                                     std::string_view bytes) {
  hash = hash_u64(hash, bytes.size());
  for (unsigned char byte : bytes) {
    hash = hash_byte(hash, byte);
  }
  return hash;
}

std::uint64_t hash_bytes(std::string_view bytes) {
  std::uint64_t hash = kFnvOffset;
  for (unsigned char byte : bytes) {
    hash = hash_byte(hash, byte);
  }
  return hash;
}

std::uint64_t content_hash(const Batch& batch) {
  std::uint64_t hash = hash_bytes_with_length(kFnvOffset, batch.label());
  hash = hash_u64(hash, batch.items_size());
  for (const auto& item : batch.items()) {
    hash = hash_u64(hash, item.id());
    hash = hash_bytes_with_length(hash, item.name());
    hash = hash_u64(hash, item.scores_size());
    for (std::int32_t score : item.scores()) {
      hash = hash_u64(hash, static_cast<std::uint32_t>(score));
    }
    hash = hash_bytes_with_length(hash, item.payload());
    if (item.has_meta()) {
      hash = hash_byte(hash, 1);
      hash = hash_bytes_with_length(hash, item.meta().source());
      hash = hash_u64(hash, item.meta().created_at());
      hash = hash_byte(hash, item.meta().active() ? 1 : 0);
    } else {
      hash = hash_byte(hash, 0);
    }
    hash = hash_u64(hash, item.tags_size());
    for (const auto& tag : item.tags()) {
      hash = hash_bytes_with_length(hash, tag);
    }
    hash = hash_bytes_with_length(hash, item.note());
  }
  return hash;
}

Batch make_batch(std::uint64_t item_count) {
  if (item_count >
      static_cast<std::uint64_t>(std::numeric_limits<int>::max())) {
    throw std::runtime_error("item count is too large for the C++ runtime");
  }

  Batch batch;
  batch.set_label("batch-" + std::to_string(item_count));
  batch.mutable_items()->Reserve(static_cast<int>(item_count));
  for (std::uint64_t i = 0; i < item_count; ++i) {
    auto* item = batch.add_items();
    item->set_id(static_cast<std::uint32_t>(i));
    item->set_name("item-" + std::to_string(i));
    for (std::uint64_t j = 0; j < 8; ++j) {
      const auto score =
          static_cast<std::int64_t>((i + 1) * (j + 3)) - 19;
      item->add_scores(static_cast<std::int32_t>(score));
    }
    std::string payload;
    payload.reserve(48 + i % 16);
    for (std::uint64_t j = 0; j < 48 + i % 16; ++j) {
      payload.push_back(
          static_cast<char>((i * 31 + j * 17 + 13) % 251));
    }
    item->set_payload(std::move(payload));
    auto* metadata = item->mutable_meta();
    metadata->set_source("source-" + std::to_string(i % 11));
    metadata->set_created_at(1700000000ULL + i * 17);
    metadata->set_active(i % 2 == 0);
    item->add_tags("tag-" + std::to_string(i % 5));
    item->add_tags("group-" + std::to_string(i % 9));
    item->add_tags("bucket-" + std::to_string(i % 13));
    item->add_tags("region-" + std::to_string(i % 7));
    item->set_note("note-" + std::to_string(i % 17) + "-" +
                   std::to_string(i * 3));
  }
  return batch;
}

std::string encode(const Batch& batch) {
  std::string bytes;
  if (!batch.SerializeToString(&bytes)) {
    throw std::runtime_error("C++ protobuf serialization failed");
  }
  return bytes;
}

Batch decode(std::string_view bytes) {
  Batch batch;
  if (!batch.ParseFromArray(bytes.data(), static_cast<int>(bytes.size()))) {
    throw std::runtime_error("C++ protobuf parsing failed");
  }
  return batch;
}

std::uint64_t consume_bytes(std::string_view bytes) {
  if (bytes.empty()) {
    return 0;
  }
  return bytes.size() + static_cast<unsigned char>(bytes.front()) +
         static_cast<unsigned char>(bytes.back());
}

std::uint64_t consume_batch(const Batch& batch) {
  if (batch.items().empty()) {
    return batch.label().size();
  }
  return batch.items_size() + batch.items(0).id() +
         batch.items(batch.items_size() - 1).id() + batch.label().size();
}

std::uint64_t elapsed_ns(Clock::time_point start, Clock::time_point stop) {
  return std::chrono::duration_cast<std::chrono::nanoseconds>(stop - start)
      .count();
}

std::uint64_t parse_u64(const char* name, const char* value) {
  try {
    std::size_t consumed = 0;
    const auto parsed = std::stoull(value, &consumed);
    if (consumed != std::string(value).size()) {
      throw std::invalid_argument("trailing content");
    }
    return parsed;
  } catch (const std::exception&) {
    throw std::runtime_error(std::string("invalid ") + name + ": " + value);
  }
}

struct Result {
  std::uint64_t data_setup_ns = 0;
  std::uint64_t input_setup_ns = 0;
  std::uint64_t first_ns = 0;
  std::uint64_t steady_ns = 0;
  std::uint64_t output_bytes = 0;
  std::uint64_t content_hash = 0;
  std::uint64_t output_hash = 0;
  std::uint64_t checksum = 0;
};

void print_result(std::string_view operation, std::uint64_t items,
                  std::uint64_t iterations, bool validate,
                  const Result& result) {
  const auto ns_per_op =
      iterations == 0 ? 0 : result.steady_ns / iterations;
  std::cout << "BENCH_RESULT implementation=cpp-binary"
            << " operation=" << operation << " items=" << items
            << " iterations=" << iterations
            << " data_setup_ns=" << result.data_setup_ns
            << " input_setup_ns=" << result.input_setup_ns
            << " first_ns=" << result.first_ns
            << " steady_ns=" << result.steady_ns
            << " steady_ns_per_op=" << ns_per_op
            << " output_bytes=" << result.output_bytes
            << " content_hash=" << result.content_hash
            << " output_hash=" << result.output_hash
            << " checksum=" << result.checksum
            << " validation=" << (validate ? 1 : 0)
            << " runtime_version=" << GOOGLE_PROTOBUF_VERSION << '\n';
}

Result run_encode(std::uint64_t items, std::uint64_t iterations,
                  bool validate) {
  Result result;
  auto start = Clock::now();
  Batch batch = make_batch(items);
  auto stop = Clock::now();
  result.data_setup_ns = elapsed_ns(start, stop);
  result.content_hash = content_hash(batch);

  start = Clock::now();
  std::string last = encode(batch);
  stop = Clock::now();
  result.first_ns = elapsed_ns(start, stop);
  result.checksum = consume_bytes(last);

  start = Clock::now();
  for (std::uint64_t i = 0; i < iterations; ++i) {
    std::string bytes = encode(batch);
    result.checksum += consume_bytes(bytes);
    last = std::move(bytes);
  }
  stop = Clock::now();
  result.steady_ns = elapsed_ns(start, stop);

  if (validate) {
    const Batch decoded = decode(last);
    const auto actual_hash = content_hash(decoded);
    if (actual_hash != result.content_hash) {
      throw std::runtime_error("cpp-binary encode content mismatch");
    }
  }
  result.output_bytes = last.size();
  result.output_hash = hash_bytes(last);
  return result;
}

Result run_decode(std::uint64_t items, std::uint64_t iterations,
                  bool validate) {
  Result result;
  auto start = Clock::now();
  Batch batch = make_batch(items);
  auto stop = Clock::now();
  result.data_setup_ns = elapsed_ns(start, stop);
  result.content_hash = content_hash(batch);

  start = Clock::now();
  const std::string input = encode(batch);
  stop = Clock::now();
  result.input_setup_ns = elapsed_ns(start, stop);

  start = Clock::now();
  Batch last = decode(input);
  stop = Clock::now();
  result.first_ns = elapsed_ns(start, stop);
  result.checksum = consume_batch(last);

  start = Clock::now();
  for (std::uint64_t i = 0; i < iterations; ++i) {
    Batch decoded = decode(input);
    result.checksum += consume_batch(decoded);
    last = std::move(decoded);
  }
  stop = Clock::now();
  result.steady_ns = elapsed_ns(start, stop);

  if (validate && content_hash(last) != result.content_hash) {
    throw std::runtime_error("cpp-binary decode content mismatch");
  }
  result.output_bytes = input.size();
  result.output_hash = hash_bytes(input);
  return result;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    if (argc == 2 && std::string_view(argv[1]) == "startup") {
      std::cout
          << "BENCH_RESULT implementation=cpp-runtime operation=startup "
             "items=0 iterations=0 data_setup_ns=0 input_setup_ns=0 first_ns=0 "
             "steady_ns=0 steady_ns_per_op=0 output_bytes=0 content_hash=0 "
             "output_hash=0 checksum=0 validation=0 runtime_version="
          << GOOGLE_PROTOBUF_VERSION << '\n';
      return 0;
    }
    if (argc == 2 && std::string_view(argv[1]) == "version") {
      std::cout << BENCH_PROTOBUF_RELEASE << " " << GOOGLE_PROTOBUF_VERSION
                << '\n';
      return 0;
    }
    if (argc != 5) {
      throw std::runtime_error(
          "usage: <encode|decode> <items> <steady-iterations> <validate:0|1>");
    }
    const std::string_view operation = argv[1];
    const auto items = parse_u64("items", argv[2]);
    const auto iterations = parse_u64("steady-iterations", argv[3]);
    const auto validate_number = parse_u64("validate", argv[4]);
    if (validate_number > 1) {
      throw std::runtime_error("validate must be 0 or 1");
    }
    const bool validate = validate_number == 1;
    if (operation == "encode") {
      print_result(operation, items, iterations, validate,
                   run_encode(items, iterations, validate));
    } else if (operation == "decode") {
      print_result(operation, items, iterations, validate,
                   run_decode(items, iterations, validate));
    } else {
      throw std::runtime_error("operation must be encode or decode");
    }
    return 0;
  } catch (const std::exception& error) {
    std::cerr << "benchmark error: " << error.what() << '\n';
    return 2;
  }
}
